src/main/java/net/neoforged/neoforge/codec/IndexedNonNullListCodec.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.codec;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.MapCodec;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import net.minecraft.core.NonNullList;
+
+/**
+ * This codec serializes an instance of {@code NonNullList<T>} and preserves list index information.
+ * It is particularly helpful in scenarios where "empty" values must be ignored, such as inventories.
+ *
+ * @param elementCodec  Codec for the elements of the list.
+ * @param defaultValue  A default value, used to fill empty lists on creation.
+ * @param skipPredicate A predicate used to skip serialization of default/empty list items.
+ * @param indexKey      A key used for serializing list index information.
+ * @param <E>           Type of elements in list.
+ */
+public record IndexedNonNullListCodec<E>(MapCodec<E> elementCodec, E defaultValue, Predicate<E> skipPredicate, String indexKey) implements Codec<NonNullList<E>> {
```

Comment:
This class is oddly specific to the old serialization pattern. Shouldn't we be moving towards always serializing entries?

Reply:
If we're willing to break old-NBT support, this could be simplified quite a bit to no longer require a MapCodec. I believe newer codecs for stuff like inventories (a prime usage of NNL) simply encode the data as a group alongside the index, rather than inlining.